File: examples/ruby/projector-log-transaction/lib/log_projector_logic.rb
# frozen_string_literal: true

require 'json'
require 'time'

# Pure business logic for transaction log projector.
# No gRPC dependencies - can be tested in isolation.
module LogProjectorLogic
  LogEntry = Struct.new(:event_type, :event_data, :is_known, keyword_init: true)

  class << self
    def process_event(event_any)
      return nil unless event_any

      type_url = event_any.type_url
      event_type = type_url.split('/').last

      if known_event?(type_url)
        event_data = decode_event(type_url, event_any.value)
        LogEntry.new(event_type: event_type, event_data: event_data, is_known: true)
      else
        LogEntry.new(event_type: event_type, event_data: { raw_type: type_url }, is_known: false)
      end
    end

    def format_log(log_entry, domain: nil, aggregate_id: nil, sequence: nil, created_at: nil)
      {
        level: 'info',
        message: log_entry.is_known ? 'transaction event' : 'unknown event',
        projector: 'projector-log-transaction',
        event_type: log_entry.event_type,
        sequence: sequence,
        domain: domain,
        aggregate_id: aggregate_id,
        created_at: created_at,
        event_data: log_entry.event_data,
        timestamp: Time.now.iso8601
      }
    end

    private

    def known_event?(type_url)
      type_url.end_with?('TransactionCreated') ||
        type_url.end_with?('DiscountApplied') ||
        type_url.end_with?('TransactionCompleted') ||
        type_url.end_with?('TransactionCancelled')
    end

    def decode_event(type_url, value)
      if type_url.end_with?('TransactionCreated')
        event = Examples::TransactionCreated.decode(value)
        { customer_id: event.customer_id, subtotal_cents: event.subtotal_cents }
      elsif type_url.end_with?('DiscountApplied')
        event = Examples::DiscountApplied.decode(value)
        { discount_type: event.discount_type, discount_cents: event.discount_cents }
      elsif type_url.end_with?('TransactionCompleted')
        event = Examples::TransactionCompleted.decode(value)
        { final_total_cents: event.final_total_cents, payment_method: event.payment_method }
      elsif type_url.end_with?('TransactionCancelled')
        event = Examples::TransactionCancelled.decode(value)
        { reason: event.reason }
      else
        { raw_type: type_url }
      end
    end
  end
end

File: examples/kotlin/projector-log-transaction/src/main/kotlin/dev/angzarr/examples/projector/LogProjectorLogic.kt
package dev.angzarr.examples.projector

import dev.angzarr.EventBook
import com.google.protobuf.Any
import examples.Domains.DiscountApplied
import examples.Domains.TransactionCancelled
import examples.Domains.TransactionCompleted
import examples.Domains.TransactionCreated

/**
 * Result of processing an event for logging.
 */
data class LogEntry(
    val eventType: String,
    val domain: String,
    val rootId: String,
    val sequence: Int,
    val fields: Map<String, String>
)

/**
 * Interface for log projector business logic.
 */
interface LogProjectorLogic {
    /**
     * Process an event book and return log entries for each event.
     *
     * @param eventBook the event book to process
     * @return list of log entries
     */
    fun processEvents(eventBook: EventBook): List<LogEntry>
}

/**
 * Default implementation of transaction log projector logic.
 */
class DefaultTransactionLogProjectorLogic : LogProjectorLogic {

    override fun processEvents(eventBook: EventBook): List<LogEntry> {
        val domain = eventBook.cover?.domain ?: "transaction"
        val rootId = eventBook.cover?.root?.value?.toByteArray()
            ?.joinToString("") { "%02x".format(it) } ?: ""

        return eventBook.pagesList.mapNotNull { page ->
            val event = page.event ?: return@mapNotNull null
            processEvent(event, domain, rootId, page.num)
        }
    }

    private fun processEvent(event: Any, domain: String, rootId: String, sequence: Int): LogEntry {
        return when {
            event.`is`(TransactionCreated::class.java) -> {
                val e = event.unpack(TransactionCreated::class.java)
                LogEntry(
                    eventType = "TransactionCreated",
                    domain = domain,
                    rootId = rootId,
                    sequence = sequence,
                    fields = mapOf(
                        "customer" to e.customerId,
                        "items" to e.itemsCount.toString(),
                        "subtotal" to e.subtotalCents.toString()
                    )
                )
            }
            event.`is`(DiscountApplied::class.java) -> {
                val e = event.unpack(DiscountApplied::class.java)
                LogEntry(
                    eventType = "DiscountApplied",
                    domain = domain,
                    rootId = rootId,
                    sequence = sequence,
                    fields = mapOf(
                        "discount_type" to e.discountType,
                        "value" to e.value.toString(),
                        "cents" to e.discountCents.toString()
                    )
                )
            }
            event.`is`(TransactionCompleted::class.java) -> {
                val e = event.unpack(TransactionCompleted::class.java)
                LogEntry(
                    eventType = "TransactionCompleted",
                    domain = domain,
                    rootId = rootId,
                    sequence = sequence,
                    fields = mapOf(
                        "total" to e.finalTotalCents.toString(),
                        "payment" to e.paymentMethod,
                        "points" to e.loyaltyPointsEarned.toString()
                    )
                )
            }
            event.`is`(TransactionCancelled::class.java) -> {
                val e = event.unpack(TransactionCancelled::class.java)
                LogEntry(
                    eventType = "TransactionCancelled",
                    domain = domain,
                    rootId = rootId,
                    sequence = sequence,
                    fields = mapOf("reason" to e.reason)
                )
            }
            else -> LogEntry(
                eventType = "unknown",
                domain = domain,
                rootId = rootId,
                sequence = sequence,
                fields = emptyMap()
            )
        }
    }
}
